
Our analysis will answer the question: ‘Does a correlation exist between fluctuating levels of dissolved oxygen, salinity, and water temperature in various waterways located in Virginia, and if so, do these correspond with population changes in marine ecosystems?’ 

Import the water quality data from this file into a pandas data frame. Then display the first few rows.

Our project will focus on the data columns relating to salinity, dissolved oxygen, and water temperature. 

In [1]:
import pandas as pd
filename = "https://raw.githubusercontent.com/ENVS110a-SP23/group8-project/main/BKB_WaterQualityData_2020084%20(1).csv"
df = pd.read_csv(filename)

df.head()



Site_Id Unit_Id  Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
0     Bay     NaN   1/3/1994             1.3                     11.7   
1     Bay     NaN  1/31/1994             1.5                     12.0   
2     Bay     NaN   2/7/1994             1.0                     10.5   
3     Bay     NaN  2/23/1994             1.0                     10.1   
4     Bay     NaN  2/28/1994             1.0                     12.6   

   pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
0                  7.3              0.40             0.40              5.9   
1                  7.4              0.20             0.35              3.0   
2                  7.2              0.25             0.60              5.9   
3                  7.4              0.35             0.50             10.0   
4                  7.2              0.20             0.40              1.6   

   Air Temp-Celsius  Air Temp (?F) Time (24:00) Field_Tech DateVerified  \
0               8.0          46.40        11:00        NaN          NaN   
1               2.6          36.68        11:30        NaN          NaN   
2               7.6          45.68         9:45        NaN          NaN   
3               2.7          36.86          NaN        NaN          NaN   
4               0.0          32.00        10:30        NaN          NaN   

  WhoVerified  AirTemp (C)  Year  
0         NaN          8.0  1994  
1         NaN          2.6  1994  
2         NaN          7.6  1994  
3         NaN          2.7  1994  
4         NaN          0.0  1994

In [2]:
df.shape

(2371, 17)

Provide summary statistics about all the numeric columns. Since the very top row counts non-missing values by column, you can tell that dissolved oxygen is missing nearly 1/3 of the data. The other columns, temperature and salinity, have more data. In the future we will write code that will analyze the means and standard deviations. 

In [3]:
df.describe()

Salinity (ppt)  Dissolved Oxygen (mg/L)  pH (standard units)  \
count     2241.000000              1520.000000          2276.000000   
mean         0.717068                 6.646263             7.168212   
std          1.230819                 2.506608             0.788485   
min          0.000000                 0.000000             0.300000   
25%          0.000000                 4.800000             6.500000   
50%          0.000000                 6.500000             7.000000   
75%          1.000000                 8.500000             7.500000   
max          9.000000                15.100000             9.900000   

       Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  Air Temp-Celsius  \
count       2298.000000      2300.000000      2250.000000         85.000000   
mean           0.524898         0.762559        18.062138         16.437647   
std            0.473663         0.621140         8.298246         11.754138   
min            0.000000         0.010000         0.000000          0.000000   
25%            0.300000         0.400000        11.000000          9.000000   
50%            0.400000         0.650000        19.000000         15.000000   
75%            0.650000         0.950000        25.000000         21.700000   
max            9.000000        12.000000        74.000000         74.000000   

       Air Temp (?F)  AirTemp (C)         Year  
count    2300.000000  2371.000000  2371.000000  
mean       62.051637    15.663050  2006.049346  
std        15.492236    10.314720     9.002373  
min        10.500000   -17.777778  1899.000000  
25%        49.000000     8.888889  2001.000000  
50%        63.000000    16.666667  2007.000000  
75%        75.000000    23.888889  2012.000000  
max        92.300000    33.500000  2019.000000

printing summary of the missing data:

In [38]:
print('dataframe shape (size): ', df.shape)
null_values = df.isnull().sum().sum()
print("Number of NaN (can start as N/A, None, or null) values: ", null_values)
# count number of Not Recorded values present in the data frame
nr_values = df.isin(['Not Recorded']).sum().sum()
print("Number of Not Recorded values: ",nr_values)

dataframe shape (size):  (2371, 17)
Number of NaN (can start as N/A, None, or null) values:  9981
Number of Not Recorded values:  1225


A new look at the earlier table of summary statistics restricted to these 6 columns. Since there are only 85 numeric values of the Air-temp-Celsius column we may not be able to use this variable. 

In [13]:
#summary statistics of key variables
df[['Salinity (ppt)', 'Dissolved Oxygen (mg/L)', 'pH (standard units)','Secchi Depth (m)','Water Depth (m)', 'Air Temp-Celsius']].describe()

Salinity (ppt)  Dissolved Oxygen (mg/L)  pH (standard units)  \
count     2241.000000              1520.000000          2276.000000   
mean         0.717068                 6.646263             7.168212   
std          1.230819                 2.506608             0.788485   
min          0.000000                 0.000000             0.300000   
25%          0.000000                 4.800000             6.500000   
50%          0.000000                 6.500000             7.000000   
75%          1.000000                 8.500000             7.500000   
max          9.000000                15.100000             9.900000   

       Secchi Depth (m)  Water Depth (m)  Air Temp-Celsius  
count       2298.000000      2300.000000         85.000000  
mean           0.524898         0.762559         16.437647  
std            0.473663         0.621140         11.754138  
min            0.000000         0.010000          0.000000  
25%            0.300000         0.400000          9.000000  
50%            0.400000         0.650000         15.000000  
75%            0.650000         0.950000         21.700000  
max            9.000000        12.000000         74.000000

locating the missing data (null):

display a row of data if there is an N/A or NaN in any column. By looking at the data this way we may be able to identify patterns of missing data, for instance data is missing in some years but not others, some locations and not others, etc. 

In [48]:
df.loc[df.isnull().any(axis=1),:].reset_index(drop=True)

Site_Id Unit_Id   Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
0        Bay     NaN    1/3/1994             1.3                     11.7   
1        Bay     NaN   1/31/1994             1.5                     12.0   
2        Bay     NaN    2/7/1994             1.0                     10.5   
3        Bay     NaN   2/23/1994             1.0                     10.1   
4        Bay     NaN   2/28/1994             1.0                     12.6   
...      ...     ...         ...             ...                      ...   
2362     Bay     NaN  10/11/2018             1.9                      5.0   
2363     Bay     NaN  10/24/2018             0.0                      9.0   
2364     Bay     NaN  10/28/2018             0.9                      2.9   
2365     Bay     NaN   11/7/2018             1.7                      NaN   
2366     Bay     NaN  12/11/2018             0.1                      NaN   

      pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
0                     7.3              0.40             0.40              5.9   
1                     7.4              0.20             0.35              3.0   
2                     7.2              0.25             0.60              5.9   
3                     7.4              0.35             0.50             10.0   
4                     7.2              0.20             0.40              1.6   
...                   ...               ...              ...              ...   
2362                  7.0              4.00             1.20             25.0   
2363                  7.0              0.30             0.60             18.0   
2364                  7.0              0.40             0.90             13.0   
2365                  7.0              0.45             0.90             20.0   
2366                  7.0              0.10             0.10             10.0   

      Air Temp-Celsius  Air Temp (?F) Time (24:00) Field_Tech DateVerified  \
0                  8.0          46.40        11:00        NaN          NaN   
1                  2.6          36.68        11:30        NaN          NaN   
2                  7.6          45.68         9:45        NaN          NaN   
3                  2.7          36.86          NaN        NaN          NaN   
4                  0.0          32.00        10:30        NaN          NaN   
...                ...            ...          ...        ...          ...   
2362               NaN          78.00        09:30    Sue Poe   11/13/2019   
2363               NaN          58.00        09:30    Sue Poe   11/13/2019   
2364               NaN          49.00        09:20    Sue Poe   11/13/2019   
2365               NaN          65.00        09:45    Sue Poe   11/13/2019   
2366               NaN          42.00        09:40    Sue Poe   11/13/2019   

          WhoVerified  AirTemp (C)  Year  
0                 NaN     8.000000  1994  
1                 NaN     2.600000  1994  
2                 NaN     7.600000  1994  
3                 NaN     2.700000  1994  
4                 NaN     0.000000  1994  
...               ...          ...   ...  
2362  Christine Folks    25.555556  2018  
2363  Christine Folks    14.444444  2018  
2364  Christine Folks     9.444444  2018  
2365  Christine Folks    18.333333  2018  
2366  Christine Folks     5.555556  2018  

[2367 rows x 17 columns]

locating the missing data (Not Recorded): 

Repeating previous command with the "Field_Tech" column because it can actually say 'Not Recorded'. 'Not Recorded' shows up 1225 times in the "Field_Tech" column. 

In [49]:
df.loc[df.isin(['Not Recorded']).any(axis=1),:].reset_index(drop=True)

Site_Id Unit_Id   Read_Date  Salinity (ppt)  Dissolved Oxygen (mg/L)  \
0          A     NaN   8/11/1999             3.0                     2.60   
1          A     NaN   8/18/1999             3.0                     1.60   
2          A     NaN    9/8/1999             1.8                     1.35   
3          A     NaN  10/15/1999             0.1                     5.65   
4          A     NaN  11/10/1999             0.0                     6.70   
...      ...     ...         ...             ...                      ...   
1220       D     NaN  10/22/2013             0.0                     7.00   
1221       D     NaN  11/12/2013             0.0                      NaN   
1222       D     NaN  11/19/2013             0.0                     9.60   
1223       D     NaN   12/4/2013             0.0                      NaN   
1224       D     NaN  12/11/2013             0.0                     8.65   

      pH (standard units)  Secchi Depth (m)  Water Depth (m)  Water Temp (?C)  \
0                     7.7              0.60             0.60             26.0   
1                     8.7               NaN              NaN             31.0   
2                     8.2              0.40             1.00             28.0   
3                     7.3              0.60             0.80             20.0   
4                     7.6              0.65             0.95             18.0   
...                   ...               ...              ...              ...   
1220                  6.5              1.00             1.20             17.0   
1221                  6.0              0.80             1.10             12.0   
1222                  6.5              1.20             1.20             12.0   
1223                  6.5              1.20             1.20             10.0   
1224                  6.5              0.80             0.80              8.0   

      Air Temp-Celsius  Air Temp (?F) Time (24:00)    Field_Tech DateVerified  \
0                  NaN          80.96        23:02  Not Recorded          NaN   
1                  NaN          91.40         9:36  Not Recorded          NaN   
2                  NaN          76.60        14:24  Not Recorded          NaN   
3                  NaN            NaN         0:00  Not Recorded          NaN   
4                  NaN            NaN         0:00  Not Recorded          NaN   
...                ...            ...          ...           ...          ...   
1220               NaN          58.00         0:00  Not Recorded          NaN   
1221               NaN          54.00         0:00  Not Recorded          NaN   
1222               NaN          52.00         0:00  Not Recorded          NaN   
1223               NaN          51.00         0:00  Not Recorded          NaN   
1224               NaN          38.00         0:00  Not Recorded          NaN   

     WhoVerified  AirTemp (C)  Year  
0            NaN    27.200000  1999  
1            NaN    33.000000  1999  
2            NaN    24.777778  1999  
3            NaN   -17.777778  1999  
4            NaN   -17.777778  1999  
...          ...          ...   ...  
1220         NaN    14.444444  2013  
1221         NaN    12.222222  2013  
1222         NaN    11.111111  2013  
1223         NaN    10.555556  2013  
1224         NaN     3.333333  2013  

[1225 rows x 17 columns]

Through our data processing we have discovered that there is a lot of missing data including in the columns we want to focus on. We will have to decide as a group how we want to navigate this because deleting columns with missing data could scew our final results. For example, if the salinity column is always missing data when the temperature column exhibits a low value (low temp.), and we decided to throw those rows of data away, we would be ignoring data about low temperatures and so our conclusions would be scewed to higher temperatures. When we present our results we would have to communicate the missing data. 